### PR TITLE
 Objects that are temporal should set _isScopedTemplate to true

### PR DESCRIPTION
**Pull request by hermanventer**
Release note: Fixes problem with result of Array.filter (and similar) applied to arrays with unknown properties

Closes: #1924

Generator.deriveConcrete can result in temporal object values. These must be marked with _isScopedTemplate to prevent them from being hoisted.

### CODE REVIEW

**Comment by gaearon** [comment]
Looks like there's a Prepack model invariant violation.

```
var inner, fn;
(function () {
  var _$6 = this;

  var _$7 = _$6.Object;
  var _$8 = _$7.assign;
  var _$9 = _$6.Array;
  var _$A = _$9.from;
  var _$B = _$6.Error;
  var _$C = _$7.prototype;
  var _$D = _$C.hasOwnProperty;
  var _$E = _$9.prototype;
  var _$F = _$E.filter;
  var _$G = _$6.Boolean;

  var _2 = function (arg) {
    var _3 = _$8(arg);

    var _$0 = _3.condition;

    var _5 = !_$0;

    var _M = _$9;
    var _N = _$A;
    var _K = _$E;
    var _L = _$F;
    var _P = _$G;

    if (!_5) {
      if (typeof _M.from !== "function") throw new _$B("Prepack model invariant violation (" + 0 + "): " + _M.from);

      var _7 = _$8(arg);

      var _$1 = _7.foo;

      var _$2 = _N(_$1);

      var _8 = _$2;
      if (_$D.call(_8, "filter")) throw new _$B("Prepack model invariant violation (" + 1 + "): " + _8.filter);
      if (typeof _K.filter !== "function") throw new _$B("Prepack model invariant violation (" + 2 + "): " + _K.filter);

      var _$3 = _8.filter(_P);

      var _9 = _$3;
      if (_$D.call(_9, "0")) throw new _$B("Prepack model invariant violation (" + 3 + "): " + _9[0]);
      var _$4 = _9["0"];
      if (_$D.call(_9, "length")) throw new _$B("Prepack model invariant violation (" + 4 + "): " + _9.length);
      var _$5 = _9["length"];
    }

    var _G = 0 === _$5;

    var _E = _G ? null : 42;

    var _C = _5 ? null : _E;

    return _C;
  };

  var _J = function (props) {
    var foo = Array.from(props.foo);
    var bar = foo.filter(Boolean);
    bar[0];

    if (bar.length === 0) {
      return null;
    }

    return 42;
  };

  var _0 = function () {
    return JSON.stringify([fn({
      condition: false
    }), fn({
      condition: true,
      foo: []
    }), fn({
      condition: true,
      foo: [false]
    }), fn({
      condition: true,
      foo: [true]
    }), fn({
      condition: true,
      foo: [false, 5]
    }), fn({
      condition: true,
      foo: [true, 5]
    })]);
  };

  _$6.inner = _J;
  _$6.fn = _2;
  var _1 = _$C;
  if (_$D.call(_1, "0")) throw new _$B("Prepack model invariant violation (" + 5 + "): " + _1[0]);
  if (_$D.call(_1, "inspect")) throw new _$B("Prepack model invariant violation (" + 6 + "): " + _1.inspect);
  inspect = _0;
}).call(this);
output of inspect() on last generated code iteration
Error: Prepack model invariant violation (4): 0
evalmachine.<anonymous>:49
      if (_$D_unique27277.call(_9_unique27277, "length")) throw new _$B_unique27277("Prepack model invariant violation (" + 4 + "): " + _9_unique27277.length);
                                                          ^

Error: Prepack model invariant violation (4): 0
```

**Comment by hermanventer** [comment]
The model invariant violation came about because the result from filter was not marked as being partially known.

**Comment by gaearon** [comment]
The test failure is because we (unexpectedly?) are able to optimize inline React components.
I'd like @trueadm to have a look before we merge.

**Comment by gaearon** [comment]
I guess making it partial causes a bailout in React logic.

**Comment by trueadm** [comment]
It's failing because it can't get `Array.prototype.map` when partial because it goes down this codepath:

https://github.com/facebook/prepack/blob/master/src/methods/properties.js#L1179

**Comment by hermanventer** [comment]
It seems that an array with a widened numeric property should not be regarded as partial then. I'll see what I can do about the invariant checker.

**Comment by trueadm** [comment]
If we make them partial, we'll need to special case a bunch of places for them so that accesses to `prototype` methods passes through. I'm not sure which is the better approach.

**Comment by hermanventer** [comment]
If I make the result from Array.prototype.filter be a proper array (by using Create.ArrayValue, as we should), then everything seems to pass.

If, however, I go one step further and  set the length property of the array to be an abstract number, then the React test suite fails.

**Comment by trueadm** [comment]
So I looked into this and it took many attempts to get this working. In essense the reason the latest changes seemed to work, was because setting an abstract length value was causing a `FatalError` which meant that the pure scope recovery was kicking in and passing the newly added tests. This also meant that all the respective React tests were failing as the new special arrays were never being applied.

I looked at every approach I could think of, including adding an abstract length value. Even if I made the abstract length work, I had to make a build function so at runtime, we could properly derive the `O.length`, the problem was that this resulted in countless serialization issues. If I created the length as a temporal value, I got the furthest, but then we also emitted the `length` and I had some tests still failing when dealing with combination of nested conditions an the fact the array is intrinsic.

I then decided to add more tests around arrays and I saw even more issues with serialization, where things just weren't being emitted at all. The limitations of what our serializer can do is becoming very apparent but I managed to find a work around. 

Instead of creating an array the conventional way and assigning an abstract length, I went back to the old way of not defining a length and instead adding in checks for the special array in places where things would break. This resulted in all the tests (including the new ones) to work properly and also gave the opportunity to tidy up some parts and add better support in other areas. Furthermore, the serialized output actually looks good without bloat.

I'm going to continue working on the special arrays in a follow up PR, as I feel this PR is now in a good place to land if people are happy with it. The follow up tasks I want to deal with are:

- handling the special arrays in loops
- handling the special arrays in array spread
- handling the special arrays in Object.assign

File: src/intrinsics/ecma262/ArrayPrototype.js [hunk 1]
```diff
@@ -980,6 +980,29 @@ export default function(realm: Realm, obj: ObjectValue): void {
     // 1. Let O be ? ToObject(this value).
     let O = To.ToObject(realm, context);
 
+    // if there is an initialValue, then the call will likely
+    // have side-effects so this is not supported for now
+    if (ArrayValue.isIntrinsicAndHasWidenedNumericProperty(O)) {
+      if (initialValue || initialValue instanceof UndefinedValue) {
```
Comment (hermanventer) [inline src/intrinsics/ecma262/ArrayPrototype.js]: I don't get this condition. Did you mean:
if (initialValue && !(initialValue instanceof UndefinedValue)) {

File: src/intrinsics/ecma262/Object.js [hunk 2]
```diff
@@ -413,30 +417,66 @@ export default function(realm: Realm): NativeFunctionValue {
...
+      // If we're in pure scope and the items are completely abstract,
+      // then create an abstract temporal with an array kind
+      if (realm.isInPureScope() && obj instanceof AbstractObjectValue) {
+        let array = ArrayValue.createTemporalWithWidenedNumericProperty(
+          realm,
+          [objectValues, obj],
+          ([methodNode, objNode]) => t.callExpression(methodNode, [objNode])
+        );
+        return array;
+      } else if (ArrayValue.isIntrinsicAndHasWidenedNumericProperty(obj)) {
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: Could/should this happen in a non pure scope?

File: src/intrinsics/ecma262/ArrayPrototype.js (hunk 1 above)
Comment (trueadm) [reply src/intrinsics/ecma262/ArrayPrototype.js]: It's for when the value doesn't exist or has been passed in as undefined.

File: src/intrinsics/ecma262/Object.js (hunk 2 above)
Comment (trueadm) [reply src/intrinsics/ecma262/Object.js]: The others will need to be changed to pure scope too.

**Comment by trueadm** [comment]
@hermanventer Feel free to add/change anything that I did in my latest commit to get this PR into a state where you're happy for it to be merged.

File: src/intrinsics/ecma262/ArrayPrototype.js (hunk 1 above)
Comment (trueadm) [reply src/intrinsics/ecma262/ArrayPrototype.js]: Sorry, you were right with your first reply. Changed applied.

**Comment by gaearon** [comment]
Here is another case that this PR is not fixing:

```js
function inner(arg) {
  var bar = Array.from(arg.unknown);
  var getBar = function() {
    return bar;
  };
  return arg.calculate(getBar);
}

function fn(arg) {
  if (!arg.condition) {
    return null;
  }
  return inner(arg);
}

if (global.__optimize) __optimize(fn);

inspect = function() {
  return JSON.stringify([
    fn({condition: false}),
    fn({
      condition: true,
      unknown: 10,
      calculate(f) { return f().length; }
    }),
    fn({
      condition: true,
      unknown: [1, 2, 3],
      calculate(f) { return f().length; }
    }),
  ])
}
```

Does that indicate a related problem?

**Comment by trueadm** [comment]
@gaearon I don't believe this is a bug introduced by this PR, but rather another deeper issue that seems relates to all the other temporal issues (ordering of how they are emitted are wrong).

**Comment by gaearon** [comment]
That sounds good to me. I'll file a separate issue then.

**Pull request closed without merging**